timeseries/src/autogluon/timeseries/trainer/model_set_builder.py

```diff
@@ -0,0 +1,247 @@
+import copy
+import logging
+import re
+from collections import defaultdict
+from typing import Any, Optional, Type, Union
+
+from autogluon.common import space
+from autogluon.core import constants
+from autogluon.timeseries.configs import HYPERPARAMETER_PRESETS
+from autogluon.timeseries.metrics import TimeSeriesScorer
+from autogluon.timeseries.models import ModelRegistry
+from autogluon.timeseries.models.abstract import AbstractTimeSeriesModel, TimeSeriesModelBase
+from autogluon.timeseries.models.multi_window import MultiWindowBacktestingModel
+from autogluon.timeseries.typing import ModelHyperparameters
+from autogluon.timeseries.utils.features import CovariateMetadata
+
+logger = logging.getLogger(__name__)
+
+
+class TrainableModelSetBuilder:
+    """Responsible for building a list of model objects, in priority order, that will be trained by the
+    Trainer."""
+
+    VALID_AG_ARGS_KEYS = {
+        "name",
+        "name_prefix",
+        "name_suffix",
+    }
+
+    def __init__(
+        self,
+        path: str,
+        freq: Optional[str],
+        prediction_length: int,
+        eval_metric: TimeSeriesScorer,
+        target: str,
+        quantile_levels: list[float],
+        covariate_metadata: CovariateMetadata,
+        multi_window: bool,
+    ):
+        self.path = path
+        self.freq = freq
+        self.prediction_length = prediction_length
+        self.eval_metric = eval_metric
+        self.target = target
+        self.quantile_levels = quantile_levels
+        self.covariate_metadata = covariate_metadata
+        self.multi_window = multi_window
+
+    def get_model_set(
+        self,
+        hyperparameters: Union[str, dict, None],
+        hyperparameter_tune: bool,
+        excluded_model_types: Optional[list[str]],
+        banned_model_names: Optional[list[str]] = None,
+    ) -> list[TimeSeriesModelBase]:
```

Comment:
It currently looks a bit odd that this method returns `TimeSeriesModelBase`, but all other checks in the class are only concerned with `AbstractTimeSeriesModel`. Let's keep everything as `AbstractTimeSeriesModel` for now and then change to `TimeSeriesModelBase` if we refactor `MultiWindowBacktestingModel` to only inherit from `TimeSeriesModelBase` or decide to include the ensemble classes in the trainable model set as well?

Reply:
makes sense but the impact is slightly larger than you'd imagine. The `core.AbstractTrainer` interface kind of enforced this. I removed `construct_model_templates` from `AbstractTrainer` and also renamed the method while I'm at it.